keps/sig-node/1664-node-ips/README.md

```diff
@@ -0,0 +1,737 @@
+should keep this high-level with a focus on what signals will be looked at to
+determine graduation.
+
+Consider the following in developing the graduation criteria for this enhancement:
+- [Maturity levels (`alpha`, `beta`, `stable`)][maturity-levels]
+- [Deprecation policy][deprecation-policy]
+
+Clearly define what graduation means by either linking to the [API doc
+definition](https://kubernetes.io/docs/concepts/overview/kubernetes-api/#api-versioning),
+or by redefining what graduation means.
+
+In general, we try to use the same stages (alpha, beta, GA), regardless how the
+functionality is accessed.
+
+[maturity-levels]: https://git.k8s.io/community/contributors/devel/sig-architecture/api_changes.md#alpha-beta-and-stable-versions
+[deprecation-policy]: https://kubernetes.io/docs/reference/using-api/deprecation-policy/
+
+Below are some examples to consider, in addition to the aforementioned [maturity levels][maturity-levels].
+
+#### Alpha -> Beta Graduation
+
+- Gather feedback from developers and surveys
+- Complete features A, B, C
+- Tests are in Testgrid and linked in KEP
+
+#### Beta -> GA Graduation
+
+- N examples of real world usage
+- N installs
+- More rigorous forms of testing e.g., downgrade tests and scalability tests
+- Allowing time for feedback
+
+**Note:** Generally we also wait at least 2 releases between beta and
+GA/stable, since there's no opportunity for user feedback, or even bug reports,
+in back-to-back releases.
+
+**For non-optional features moving to GA, the graduation criteria must include [conformance tests].**
+
+[conformance tests]: https://git.k8s.io/community/contributors/devel/sig-architecture/conformance-tests.md
+-->
+
+TODO
+
+### Upgrade / Downgrade Strategy
+
+#### `--node-ip` deprecation
+
+This can happen in the usual way for deprecating command-line
+arguments.
+
+#### IPv6 `Node.Status.Addresses` Gotchas
+
+The proposed behavior has one upgrade caveat:
+
+  - If you currently have a single-stack IPv4 cluster,
+  - _and_ aren't passing any `--node-ip` value to kubelet,
+  - _and_ have software in/around the cluster that will fail if it
+    sees an IPv6 address in `Node.Status.Addresses` even when that
+    address comes after the Primary Node IP.
+  - _and_ are running on a cloud that supports IPv6
+  - _and_ are running on hosts that have IPv6 addresses assigned to
+    them even though you weren't using them (at least for Kubernetes),
+  - then you will need to update your kubelet configuration to include
+    `--node-ips ipv4` to force kubelet to be single-stack-IPv4-only in
+    the future.
+
+This should hopefully be uncommon enough to be essentially
+non-existent. And also, this does not _technically_ constitute an API
+break, since `Node.Status.Addresses` was always theoretically
+dual-stack.
+
+The alternative to this would be to make it so that if you pass no
+`--node-ips` you always get a single-stack IPv4 cluster, even in a
```

Comment:
always v4 or single-v6 would still be an option?

Reply:
Yes, what I meant here was that instead of defaulting to `ipv4,ipv6` when there was no `--node-ips`, we could default to `ipv4` instead ("single-stack IPv4 or bust"). If you wanted a single-stack IPv6 cluster you'd have to pass `--node-ips ipv6` (or `--node-ips <SPECIFIC IP>`), and if you wanted dual-stack you would have to do `--node-ips ipv4,ipv6` / `--node-ips <IP>,<IP>`.

But I feel like in the future, if you're on a node with dual-stack IPs, kubelet should default to acknowledging both of those IPs, so I didn't like this option.

Reply:
Does it matter for K8s if the node has dual addresses in a single-stack cluster?

PODs can have dual addresses since 1.9 in single-stack clusters, and IMHO it should be OK also for nodes.

Reply:
> PODs can have dual addresses since 1.9 in single-stack clusters

What do you mean? It can have multiple addresses on its "eth0" but only one of those addresses will be visible in the Kube API. (Even if the CNI plugin returns multiple addresses, kubelet will only use one of them if the dual-stack feature gate is disabled.)

And even then, the pod is only likely to get multiple addresses if the administrator has explicitly configured the network plugin to do that. The question here is about whether bad things might happen if we caused dual-stack node IPs to be exposed in the API when the admin _hadn't_ intended it.

Reply:
Ok, I see. I only considered the K8s view.